{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d2c0647d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pickle\n",
    "import networkx as nx\n",
    "from subpred.util import save_data\n",
    "from subpred.fasta import read_fasta\n",
    "import re\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bfcb6f3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASET_FOLDER= \"../data/datasets/\"\n",
    "RAW_DATA_FOLDER= \"../data/raw/\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b1b1704",
   "metadata": {},
   "source": [
    "## Uniprot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "74c64ea4",
   "metadata": {},
   "outputs": [],
   "source": [
    "names = [\n",
    "    \"Uniprot\",\n",
    "    \"gene_names\",\n",
    "    \"protein_names\",\n",
    "    \"reviewed\",\n",
    "    \"protein_existence\",\n",
    "    \"sequence\",\n",
    "    \"organism_id\",\n",
    "    \"go_ids\",\n",
    "    \"keyword_ids\",\n",
    "    \"keywords\",\n",
    "    \"tcdb_ids\",\n",
    "    \"interpro_ids\",\n",
    "]\n",
    "dtypes = {\n",
    "    \"Uniprot\": \"string\",\n",
    "    \"gene_names\": \"string\",\n",
    "    \"protein_names\": \"string\",\n",
    "    \"reviewed\": \"category\",\n",
    "    \"protein_existence\": \"category\",\n",
    "    \"sequence\": \"string\",\n",
    "    \"organism_id\": \"int\",\n",
    "    \"go_ids\": \"string\",\n",
    "    \"keyword_ids\": \"string\",\n",
    "    \"keywords\": \"string\",\n",
    "    \"tcdb_ids\": \"string\",\n",
    "    \"interpro_ids\": \"string\",\n",
    "}\n",
    "\n",
    "df = pd.read_table(\n",
    "    RAW_DATA_FOLDER+\"uniprot/uniprot_evidence1-2_nofragments.tsv\",\n",
    "    index_col=0,\n",
    "    header=None,\n",
    "    names=names,\n",
    "    dtype=dtypes,\n",
    "    skiprows=1,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "dbded4d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.reviewed = df.reviewed.transform(lambda x: x == \"reviewed\").astype(\"bool\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c40d397d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.protein_existence = df.protein_existence.map(\n",
    "    {\"Evidence at transcript level\": 2, \"Evidence at protein level\": 1}\n",
    ").astype(\"int\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f349e12d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sequences_nonstandard_aa_count = (\n",
    "#     df[~df.sequence.str.fullmatch(re.compile(\"[ACDEFGHIKLMNPQRSTVWY]+\"))]\n",
    "#     .sequence.drop_duplicates()\n",
    "#     .shape[0]\n",
    "# )\n",
    "# sequences_count = df.sequence.drop_duplicates().shape[0]\n",
    "# print(\n",
    "#     f\"proteins with non-standard amino acids: {sequences_nonstandard_aa_count} out of {sequences_count} ({sequences_nonstandard_aa_count / sequences_count * 100:.2f}%)\"\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6297b4a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# saving accessions of those proteins, for later stats\n",
    "\n",
    "df_sequences_nonstandard = df[\n",
    "    ~df.sequence.str.fullmatch(re.compile(\"[ACDEFGHIKLMNPQRSTVWY]+\"))\n",
    "].sequence.to_frame().drop_duplicates()\n",
    "save_data(df_sequences_nonstandard, \"proteins_nonstandard_aminoacids\", folder_path=DATASET_FOLDER)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d441fdb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# removing the AAs (all features are based on 20 AAs)\n",
    "df.sequence = df.sequence.str.replace(\n",
    "    re.compile(\"[^ACDEFGHIKLMNPQRSTVWY]+\"), \"\", regex=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "92300eb3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_names</th>\n",
       "      <th>protein_names</th>\n",
       "      <th>reviewed</th>\n",
       "      <th>protein_existence</th>\n",
       "      <th>sequence</th>\n",
       "      <th>organism_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uniprot</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A0A0C5B5G6</th>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>Mitochondrial-derived peptide MOTS-c (Mitochon...</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>MRWQEMGYIFYPRKLR</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0A1B0GTW7</th>\n",
       "      <td>CIROP LMLN2</td>\n",
       "      <td>Ciliated left-right organizer metallopeptidase...</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>MLLLLLLLLLLPPLVLRVAASRCLHDETQKSVSLLRPPFSQLPSKS...</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0JNW5</th>\n",
       "      <td>BLTP3B KIAA0701 SHIP164 UHRF1BP1L</td>\n",
       "      <td>Bridge-like lipid transfer protein family memb...</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>MAGIIKKQILKHLSRFTKNLSPDKINLSTLKGEGELKNLELDEEVL...</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0JP26</th>\n",
       "      <td>POTEB3</td>\n",
       "      <td>POTE ankyrin domain family member B3</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>MVAEVCSMPAASAVKKPFDLRSKMGKWCHHRFPCCRGSGKSNMGTS...</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0PK11</th>\n",
       "      <td>CLRN2</td>\n",
       "      <td>Clarin-2</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>MPGWFKKAWYGLASLLSFSSFILIIVALVVPHWLSGKILCQTGVDL...</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   gene_names  \\\n",
       "Uniprot                                         \n",
       "A0A0C5B5G6                            MT-RNR1   \n",
       "A0A1B0GTW7                        CIROP LMLN2   \n",
       "A0JNW5      BLTP3B KIAA0701 SHIP164 UHRF1BP1L   \n",
       "A0JP26                                 POTEB3   \n",
       "A0PK11                                  CLRN2   \n",
       "\n",
       "                                                protein_names  reviewed  \\\n",
       "Uniprot                                                                   \n",
       "A0A0C5B5G6  Mitochondrial-derived peptide MOTS-c (Mitochon...      True   \n",
       "A0A1B0GTW7  Ciliated left-right organizer metallopeptidase...      True   \n",
       "A0JNW5      Bridge-like lipid transfer protein family memb...      True   \n",
       "A0JP26                   POTE ankyrin domain family member B3      True   \n",
       "A0PK11                                               Clarin-2      True   \n",
       "\n",
       "            protein_existence  \\\n",
       "Uniprot                         \n",
       "A0A0C5B5G6                  1   \n",
       "A0A1B0GTW7                  1   \n",
       "A0JNW5                      1   \n",
       "A0JP26                      1   \n",
       "A0PK11                      1   \n",
       "\n",
       "                                                     sequence  organism_id  \n",
       "Uniprot                                                                     \n",
       "A0A0C5B5G6                                   MRWQEMGYIFYPRKLR         9606  \n",
       "A0A1B0GTW7  MLLLLLLLLLLPPLVLRVAASRCLHDETQKSVSLLRPPFSQLPSKS...         9606  \n",
       "A0JNW5      MAGIIKKQILKHLSRFTKNLSPDKINLSTLKGEGELKNLELDEEVL...         9606  \n",
       "A0JP26      MVAEVCSMPAASAVKKPFDLRSKMGKWCHHRFPCCRGSGKSNMGTS...         9606  \n",
       "A0PK11      MPGWFKKAWYGLASLLSFSSFILIIVALVVPHWLSGKILCQTGVDL...         9606  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sequences = df.drop(\n",
    "    [\"go_ids\", \"keyword_ids\", \"keywords\", \"tcdb_ids\", \"interpro_ids\"], axis=1\n",
    ")\n",
    "save_data(df_sequences, \"uniprot\", DATASET_FOLDER)\n",
    "df_sequences.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "23da0b37",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot</th>\n",
       "      <th>keyword</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>DNA-binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>Mitochondrion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>Nucleus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>Osteogenesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>Reference proteome</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Uniprot             keyword\n",
       "0  A0A0C5B5G6         DNA-binding\n",
       "1  A0A0C5B5G6       Mitochondrion\n",
       "2  A0A0C5B5G6             Nucleus\n",
       "3  A0A0C5B5G6        Osteogenesis\n",
       "4  A0A0C5B5G6  Reference proteome"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_keywords = (\n",
    "    df.keywords.dropna()\n",
    "    .str.split(\";\")\n",
    "    .explode()\n",
    "    .str.strip()\n",
    "    .astype(\"category\")\n",
    "    .rename(\"keyword\")\n",
    "    .to_frame()\n",
    "    .reset_index(drop=False)\n",
    "    .drop_duplicates()\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "save_data(df_keywords, \"keywords\", DATASET_FOLDER)\n",
    "df_keywords.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "fa6bc5be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot</th>\n",
       "      <th>interpro_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>IPR054130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A1B0GTW7</td>\n",
       "      <td>IPR001577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0JNW5</td>\n",
       "      <td>IPR026728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0JNW5</td>\n",
       "      <td>IPR026854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0JP26</td>\n",
       "      <td>IPR050657</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Uniprot interpro_id\n",
       "0  A0A0C5B5G6   IPR054130\n",
       "1  A0A1B0GTW7   IPR001577\n",
       "2      A0JNW5   IPR026728\n",
       "3      A0JNW5   IPR026854\n",
       "4      A0JP26   IPR050657"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_interpro = (\n",
    "    df.interpro_ids.dropna()\n",
    "    .str.rstrip(\";\")\n",
    "    .str.split(\";\")\n",
    "    .explode()\n",
    "    .rename(\"interpro_id\")\n",
    "    .to_frame()\n",
    "    .reset_index(drop=False)\n",
    "    .drop_duplicates()\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "df_interpro.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "739556c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IPR000126</th>\n",
       "      <td>Active_site</td>\n",
       "      <td>Serine proteases, V8 family, serine active site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IPR000138</th>\n",
       "      <td>Active_site</td>\n",
       "      <td>Hydroxymethylglutaryl-CoA lyase, active site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IPR000169</th>\n",
       "      <td>Active_site</td>\n",
       "      <td>Cysteine peptidase, cysteine active site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IPR000180</th>\n",
       "      <td>Active_site</td>\n",
       "      <td>Membrane dipeptidase, active site</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IPR000189</th>\n",
       "      <td>Active_site</td>\n",
       "      <td>Prokaryotic transglycosylase, active site</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  type                                             name\n",
       "accession                                                              \n",
       "IPR000126  Active_site  Serine proteases, V8 family, serine active site\n",
       "IPR000138  Active_site     Hydroxymethylglutaryl-CoA lyase, active site\n",
       "IPR000169  Active_site         Cysteine peptidase, cysteine active site\n",
       "IPR000180  Active_site                Membrane dipeptidase, active site\n",
       "IPR000189  Active_site        Prokaryotic transglycosylase, active site"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_interpro_annotations = pd.read_table(\n",
    "    RAW_DATA_FOLDER+\"interpro/interpro_entries.tsv\",\n",
    "    names=[\"accession\", \"type\", \"name\"],\n",
    "    skiprows=1,\n",
    "    index_col=0,\n",
    "    dtype={\"accession\": str, \"type\": \"category\", \"name\": str},\n",
    ")\n",
    "df_interpro_annotations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9f2d511d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot</th>\n",
       "      <th>interpro_id</th>\n",
       "      <th>type</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>IPR054130</td>\n",
       "      <td>Family</td>\n",
       "      <td>Mitochondrial-derived peptide MOTS-c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A1B0GTW7</td>\n",
       "      <td>IPR001577</td>\n",
       "      <td>Family</td>\n",
       "      <td>Peptidase M8, leishmanolysin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0JNW5</td>\n",
       "      <td>IPR026728</td>\n",
       "      <td>Family</td>\n",
       "      <td>Bridge-like lipid transfer protein family memb...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0JNW5</td>\n",
       "      <td>IPR026854</td>\n",
       "      <td>Domain</td>\n",
       "      <td>Vacuolar protein sorting-associated protein 13...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0JP26</td>\n",
       "      <td>IPR050657</td>\n",
       "      <td>Family</td>\n",
       "      <td>Ankyrin repeat domain-containing protein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Uniprot interpro_id    type  \\\n",
       "0  A0A0C5B5G6   IPR054130  Family   \n",
       "1  A0A1B0GTW7   IPR001577  Family   \n",
       "2      A0JNW5   IPR026728  Family   \n",
       "3      A0JNW5   IPR026854  Domain   \n",
       "4      A0JP26   IPR050657  Family   \n",
       "\n",
       "                                                name  \n",
       "0               Mitochondrial-derived peptide MOTS-c  \n",
       "1                       Peptidase M8, leishmanolysin  \n",
       "2  Bridge-like lipid transfer protein family memb...  \n",
       "3  Vacuolar protein sorting-associated protein 13...  \n",
       "4           Ankyrin repeat domain-containing protein  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_interpro = df_interpro.merge(\n",
    "    df_interpro_annotations, how=\"left\", left_on=\"interpro_id\", right_index=True\n",
    ").drop_duplicates()\n",
    "save_data(df_interpro, \"interpro\", DATASET_FOLDER)\n",
    "df_interpro.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2af01118",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tcdb_id</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.R.2.1.1</td>\n",
       "      <td>A0JNW5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9.A.46.1.2</td>\n",
       "      <td>A0PK11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9.B.438.1.1</td>\n",
       "      <td>A2RU14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.N.2.1.1</td>\n",
       "      <td>A6NI61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9.A.80.1.1</td>\n",
       "      <td>A6NKB5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       tcdb_id   value\n",
       "0    1.R.2.1.1  A0JNW5\n",
       "1   9.A.46.1.2  A0PK11\n",
       "2  9.B.438.1.1  A2RU14\n",
       "3    1.N.2.1.1  A6NI61\n",
       "4   9.A.80.1.1  A6NKB5"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tcdb_uniprot = (\n",
    "    df.tcdb_ids.dropna()\n",
    "    .str.rstrip(\";\")\n",
    "    .str.split(\";\")\n",
    "    .explode()\n",
    "    .rename(\"tcdb_id\")\n",
    "    .reset_index(drop=False)\n",
    "    .drop_duplicates()\n",
    "    .melt(id_vars=[\"tcdb_id\"], value_vars=[\"Uniprot\"])\n",
    ")\n",
    "df_tcdb_uniprot = df_tcdb_uniprot[[\"tcdb_id\", \"value\"]].drop_duplicates().reset_index(\n",
    "    drop=True\n",
    ")\n",
    "save_data(df_tcdb_uniprot, \"tcdb\", DATASET_FOLDER)\n",
    "df_tcdb_uniprot.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1812f7c3",
   "metadata": {},
   "source": [
    "## 3Di"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3cf8d79d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_fasta(fasta_data_3di: list, lowercase: bool = True):\n",
    "    accession_pattern = re.compile(\n",
    "        \"[OPQ][0-9][A-Z0-9]{3}[0-9]|[A-NR-Z][0-9]([A-Z][A-Z0-9]{2}[0-9]){1,2}\"\n",
    "    )\n",
    "    not_in_alphabet_pattern = re.compile(\"[^ACDEFGHIKLMNPQRSTVWYX]+\")\n",
    "    for header, sequence in fasta_data_3di:\n",
    "        header_values = header[1:].split(\"-\")\n",
    "        accession = header_values[1]\n",
    "        # number of the split, if more than 2700AA:\n",
    "        fold_number = int(header_values[2][1:])\n",
    "        if not re.fullmatch(accession_pattern, accession):\n",
    "            print(\n",
    "                \"found invalid header in 3di fasta file, unable to read accession: \"\n",
    "                + header\n",
    "            )\n",
    "            continue\n",
    "\n",
    "        if re.search(not_in_alphabet_pattern, sequence):\n",
    "            print(\"Invalid characer found in 3di sequence \" + sequence)\n",
    "            continue\n",
    "\n",
    "        # compatibility with prostt5\n",
    "        # print(sequence_lower)\n",
    "        yield accession, fold_number, sequence.lower() if lowercase else sequence\n",
    "\n",
    "\n",
    "def merge_subsequences(\n",
    "    df_uniprot_3Di: pd.DataFrame, get_stats: bool = True, overlap_len: int = 1200\n",
    "):\n",
    "    # In the case of proteins longer than 2700 amino acids (aa), AlphaFold provides 1400aa long, overlapping fragments.\n",
    "    # For example, Titin has predicted fragment structures named as Q8WZ42-F1 (residues 1–1400), Q8WZ42-F2 (residues 201–1600), etc.\n",
    "    # These fragments are currently only available for the human proteome in these proteome archive files, not on the website.\n",
    "    # All structures seem to have overlap of 1200, might change in future versions\n",
    "    df_uniprot_3Di_copy = df_uniprot_3Di.copy(deep=True)\n",
    "    # proteins that were split into more than one structure\n",
    "    split_proteins = df_uniprot_3Di_copy[\n",
    "        df_uniprot_3Di_copy.fold_number == 2\n",
    "    ].Uniprot.unique()\n",
    "\n",
    "    records_merged_sequences = list()\n",
    "    records_merging_stats = list()\n",
    "\n",
    "    for split_protein in split_proteins:\n",
    "        df_uniprot_3Di_split_protein = df_uniprot_3Di_copy[\n",
    "            df_uniprot_3Di_copy.Uniprot == split_protein\n",
    "        ][[\"fold_number\", \"sequence3Di\"]].sort_values(\"fold_number\")\n",
    "        subsequences_list = list(\n",
    "            df_uniprot_3Di_split_protein.sequence3Di.values.ravel()\n",
    "        )\n",
    "        sequence_merged = subsequences_list[0]\n",
    "        sequence_identity_scores = list()\n",
    "        for subsequence in subsequences_list[1:]:\n",
    "            matching_pos = np.equal(\n",
    "                list(subsequence[:overlap_len]), list(sequence_merged[-overlap_len:])\n",
    "            )\n",
    "            overlap_seqence_identity = sum(matching_pos) / len(matching_pos)\n",
    "            sequence_identity_scores.append(overlap_seqence_identity)\n",
    "            additional_positions = subsequence[overlap_len : len(subsequence)]\n",
    "            sequence_merged += additional_positions\n",
    "\n",
    "        records_merging_stats.append(\n",
    "            [\n",
    "                split_protein,\n",
    "                sequence_identity_scores,\n",
    "            ]\n",
    "        )\n",
    "\n",
    "        records_merged_sequences.append([split_protein, sequence_merged])\n",
    "\n",
    "    df_merged = pd.DataFrame.from_records(\n",
    "        records_merged_sequences, columns=[\"Uniprot\", \"sequence3Di\"]\n",
    "    )\n",
    "    df_uniprot_3Di_copy = df_uniprot_3Di_copy[\n",
    "        ~df_uniprot_3Di_copy.Uniprot.isin(df_merged.Uniprot)\n",
    "    ]\n",
    "    # any unmerged remain?\n",
    "    assert (df_uniprot_3Di_copy.fold_number == 1).all()\n",
    "    df_uniprot_3Di_copy = df_uniprot_3Di_copy.drop(\"fold_number\", axis=1)\n",
    "\n",
    "    df_uniprot_3Di_copy = pd.concat(\n",
    "        [df_merged, df_uniprot_3Di_copy], ignore_index=True\n",
    "    ).set_index(\"Uniprot\")\n",
    "    df_uniprot_3Di_merging_stats = pd.DataFrame.from_records(\n",
    "        records_merging_stats,\n",
    "        columns=[\n",
    "            \"Uniprot\",\n",
    "            \"merging_overlaps_seq_ident\",\n",
    "        ],\n",
    "        index=\"Uniprot\",\n",
    "    )\n",
    "    if get_stats:\n",
    "        return df_uniprot_3Di_copy, df_uniprot_3Di_merging_stats\n",
    "    else:\n",
    "        return df_uniprot_3Di_copy\n",
    "\n",
    "\n",
    "def preprocess_3di_fasta(fasta_path: str, accessions_whitelist: list):\n",
    "\n",
    "    fasta_data_3di = read_fasta(\n",
    "        fasta_path\n",
    "    )  # only contains four model organisms for now\n",
    "\n",
    "    df_uniprot_3Di = pd.DataFrame.from_records(\n",
    "        process_fasta(fasta_data_3di=fasta_data_3di),\n",
    "        columns=[\"Uniprot\", \"fold_number\", \"sequence3Di\"],\n",
    "    ).sort_values(\n",
    "        [\"Uniprot\", \"fold_number\"]\n",
    "    )  # , index=\"Uniprot\"\n",
    "\n",
    "    # Only keep those that are in our sequence dataset!\n",
    "    df_uniprot_3Di = df_uniprot_3Di[df_uniprot_3Di.Uniprot.isin(accessions_whitelist)]\n",
    "\n",
    "    # also returns stats df with sequence identity between merged structures\n",
    "    df_uniprot_3Di_merged, df_uniprot_3Di_merging_stats = merge_subsequences(\n",
    "        df_uniprot_3Di=df_uniprot_3Di\n",
    "    )\n",
    "\n",
    "    df_uniprot_3Di_merged = df_uniprot_3Di_merged.assign(\n",
    "        len_matches_aa_sequence=df.sequence.loc[df_uniprot_3Di_merged.index]\n",
    "        .str.len()\n",
    "        .values\n",
    "        == df_uniprot_3Di_merged.sequence3Di.str.len().values\n",
    "    )\n",
    "    assert ~df_uniprot_3Di_merged.index.duplicated().any()\n",
    "    return df_uniprot_3Di_merged, df_uniprot_3Di_merging_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6037432d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence3Di</th>\n",
       "      <th>len_matches_aa_sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uniprot</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A2VEC9</th>\n",
       "      <td>dddpvvvvpppppdapaakdkdkdkdkdwafpfdkdkdkdflvcvv...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4UGR9</th>\n",
       "      <td>dddddddddddddddddddddddddddddddddddddddddddpvv...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O14686</th>\n",
       "      <td>dddadddddddddddddddddddddddddddddddddddddddddd...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O15018</th>\n",
       "      <td>dfpdlvclvlvlvllvvvlvvvvvvddpdplnvlsvllsvvsnvvs...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O15050</th>\n",
       "      <td>dddppppddqllvqlvvlqvqlvvcvvvvnlvsslvslvvscvspv...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               sequence3Di  \\\n",
       "Uniprot                                                      \n",
       "A2VEC9   dddpvvvvpppppdapaakdkdkdkdkdwafpfdkdkdkdflvcvv...   \n",
       "A4UGR9   dddddddddddddddddddddddddddddddddddddddddddpvv...   \n",
       "O14686   dddadddddddddddddddddddddddddddddddddddddddddd...   \n",
       "O15018   dfpdlvclvlvlvllvvvlvvvvvvddpdplnvlsvllsvvsnvvs...   \n",
       "O15050   dddppppddqllvqlvvlqvqlvvcvvvvnlvsslvslvvscvspv...   \n",
       "\n",
       "         len_matches_aa_sequence  \n",
       "Uniprot                           \n",
       "A2VEC9                      True  \n",
       "A4UGR9                      True  \n",
       "O14686                      True  \n",
       "O15018                      True  \n",
       "O15050                      True  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_uniprot_3Di, df_uniprot_3Di_merging_stats = preprocess_3di_fasta(\n",
    "    RAW_DATA_FOLDER + \"alphafolddb/3Di_sequences.fasta\",\n",
    "    accessions_whitelist=df.index.tolist(),\n",
    ")\n",
    "median_merging_overlap = df_uniprot_3Di_merging_stats.merging_overlaps_seq_ident.apply(\n",
    "    np.median\n",
    ").median()\n",
    "mean_merging_overlap = df_uniprot_3Di_merging_stats.merging_overlaps_seq_ident.apply(\n",
    "    np.mean\n",
    ")\n",
    "save_data(df_uniprot_3Di, \"3Di_alphafold4\", folder_path=DATASET_FOLDER)\n",
    "df_uniprot_3Di.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d31d21d9",
   "metadata": {},
   "source": [
    "## GOA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7f021251",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering GO annotations... done\n"
     ]
    }
   ],
   "source": [
    "uniprot_accessions = set(df.index.unique())\n",
    "\n",
    "df_go_ebi = pd.DataFrame()\n",
    "\n",
    "print(\"Filtering GO annotations...\", end=\" \")\n",
    "with pd.read_table(\n",
    "    RAW_DATA_FOLDER+\"goa/goa_uniprot_all_ebi_filtered.tsv.gz\",\n",
    "    header=None,\n",
    "    names=[\"Uniprot\", \"qualifier\", \"go_id\", \"evidence_code\", \"aspect\", \"date\"],\n",
    "    dtype={\n",
    "        \"Uniprot\": \"string\",\n",
    "        \"qualifier\": \"category\",\n",
    "        \"go_id\": \"string\",\n",
    "        \"evidence_code\": \"category\",\n",
    "        \"aspect\": \"category\",\n",
    "        \"date\": \"int\",\n",
    "    },\n",
    "    chunksize=10**6,\n",
    "    # parse_dates=[\"date\"],\n",
    ") as reader:\n",
    "    for df_chunk in reader:\n",
    "        df_chunk_filtered = (\n",
    "            df_chunk[df_chunk.Uniprot.isin(uniprot_accessions)]\n",
    "            .drop(\"date\", axis=1)\n",
    "            .drop_duplicates()\n",
    "        )\n",
    "        df_go_ebi = pd.concat([df_go_ebi, df_chunk_filtered], axis=0).reset_index(\n",
    "            drop=True\n",
    "        )\n",
    "print(\"done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e131126b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_go_ebi = df_go_ebi.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6da86df8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using numbers starting with 0 as index halves the file size!\n",
    "df_go_ebi = df_go_ebi.reset_index(drop=True)\n",
    "\n",
    "save_data(df_go_ebi, \"goa\", DATASET_FOLDER)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c672e48c",
   "metadata": {},
   "source": [
    "## GO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87965cc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import obonet\n",
    "graph_go = obonet.read_obo(RAW_DATA_FOLDER+\"ontologies/go.obo\")\n",
    "\n",
    "save_data(graph_go, \"go_obo\", folder_path=DATASET_FOLDER)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "subpred_deeplearning",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
